features/environment.py

```diff
@@ -102,36 +103,31 @@ def reduce_db_outcomes(feature_outcomes):
                         break
 
         outcomes_to_save = list(reduce_db_outcomes(context.gherkin_outcomes))
-
-        if outcomes_to_save and context.validation_task_id is not None:
-            retrieved_task = ValidationTask.objects.get(id=context.validation_task_id)
-            retrieved_model = retrieved_task.request.model
-            retrieved_model_id = retrieved_model.id
-
-            def get_or_create_instance_when_set(spf_id):
-                if not spf_id:
-                    return None
-                # @todo see if we can change this into a bulk insert as well
-                # with bulk_create(ignore_conflicts=True). There appear to be
-                # quite some caveats regarding this though...
-                instance, _created = ModelInstance.objects.get_or_create(
-                    stepfile_id=spf_id,
-                    model_id=retrieved_model_id
-                )
-                return instance.id
-
-            spf_ids = sorted(set(o.instance_id for o in outcomes_to_save))
-            instances_in_db = list(map(get_or_create_instance_when_set, spf_ids))
-            inst_id_mapping = dict(zip(spf_ids, instances_in_db))
-
-            for outcome in outcomes_to_save:
-                # Previously we have the current model SPF id in the instance_id
-                # field. This needs to be updated to an actual foreign key into
-                # our ModelInstances table.
-                if outcome.instance_id:
-                    outcome.instance_id = inst_id_mapping[outcome.instance_id]
-
-            ValidationOutcome.objects.bulk_create(outcomes_to_save)
+        outcomes_instances_to_save = list()
+
+        if outcomes_to_save:
+            with transaction.atomic():
+                task = ValidationTask.objects.get(id=context.validation_task_id)
+                model_id = task.request.model.id
+
+                stepfile_ids = sorted(set(o.instance_id for o in outcomes_to_save))
+                for stepfile_id in stepfile_ids:
+                    if stepfile_id:
+                        instance = ModelInstance(
+                            stepfile_id=stepfile_id,
+                            model_id=model_id
+                        )
+                        outcomes_instances_to_save.append(instance)
```

Comment:
```suggestion
                  instance = ModelInstance(
                      stepfile_id=stepfile_id,
                      model_id=model_id
                  )
                  outcomes_instances_to_save.append(instance)
```

Then the if is no longer needed here